cilium: completion of nat46/64 gateway

borkmann opened the pull request:
(see commit msgs)

borkmann committed:
bpf: Add generic stateful nat6->4 support

This work completes the standalone egress GW functionality by implementing
generic nat64 without having to configure specific services for remap:

On the Gateway node:

  # ./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=true \
    --datapath-mode=lb-only --bpf-lb-algorithm=maglev \
    --bpf-lb-maglev-table-size=2039 --bpf-lb-acceleration=disabled \
    --devices=enp5s0 --disable-envoy-version-check=true \
    --enable-stateless-nat46x64=true

  # ./cilium/cilium service list
  ID   Frontend   Service Type   Backend
  #

  # tcpdump -i enp5s0 port 80 -n
  [...]
  12:06:05.823733 IP6 2a02:168:f656:0:1ac0:4dff:ff01:d5e6.54054 > 64:ff9b::8c52:7903.80: Flags [S], seq 2743947568, win 43200, options [mss 1440,sackOK,TS val 3724070640 ecr 0,nop,wscale 9], length 0
  12:06:05.823745 IP 192.168.2.11.54054 > 140.82.121.3.80: Flags [S], seq 2743947568, win 43200, options [mss 1440,sackOK,TS val 3724070640 ecr 0,nop,wscale 9], length 0
  12:06:05.830454 IP 140.82.121.3.80 > 192.168.2.11.54054: Flags [S.], seq 433570079, ack 2743947569, win 65535, options [mss 1436,sackOK,TS val 1565141015 ecr 3724070640,nop,wscale 10], length 0
  12:06:05.830479 IP6 64:ff9b::8c52:7903.80 > 2a02:168:f656:0:1ac0:4dff:ff01:d5e6.54054: Flags [S.], seq 433570079, ack 2743947569, win 65535, options [mss 1436,sackOK,TS val 1565141015 ecr 3724070640,nop,wscale 10], length 0
  [...]

The client only needs to add a next hop entry in its routing table, no
special source address selection is needed:

  # ip -6 r
  [...]
  64:ff9b::/96 via 2a02:168:f656:0:1ac0:4dff:ff01:c164 dev enp5s0 metric 1024 pref medium
  [...]

  # curl --verbose "http://[64:ff9b::101:101]"
  *   Trying 64:ff9b::101:101:80...
  * TCP_NODELAY set
  * Connected to 64:ff9b::101:101 (64:ff9b::101:101) port 80 (#0)
  > GET / HTTP/1.1
  > Host: [64:ff9b::101:101]
  > User-Agent: curl/7.68.0
  > Accept: */*
  >
  * Mark bundle as not supporting multiuse
  < HTTP/1.1 403 Forbidden
  < Date: Tue, 29 Nov 2022 13:30:30 GMT
  < Content-Type: text/plain; charset=UTF-8
  < Content-Length: 16
  < Connection: close
  < X-Frame-Options: SAMEORIGIN
  < Referrer-Policy: same-origin
  < Cache-Control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0
  < Expires: Thu, 01 Jan 1970 00:00:01 GMT
  < Server: cloudflare
  < CF-RAY: 771bb2c739d5cc3e-ZRH
  <
  * Closing connection 0
  [...]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Rename agent setting to generic nat64x64 gateway

Rename the agent setting --enable-stateless-nat46x64 into a more generic
--enable-nat46x64-gateway, given the plan is to enable both stateless and
statefull translation in this case as both can be supported in parallel.
This has not seen a stable release yet, so rename is fine.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: Extend status API to reflect service/gateway nat46x64 status

For cilium status dump we want to reflect the state of services vs
gateway support on NAT46/64. Therefore extend both into objects which
can be extended further if needed. Also for the GW we expose the current
fixed default prefix. In future, this may be a list of prefixes.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Add auto-generated openapi code for service/gateway nat46x64 status

Auto-generated API code in here.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, status: Implement verbose dump for service/gateway nat46x64 status

Add status dump and separate both Service/Gateway. Also dump the prefix for
the gateway. It's hard coded here, but subsequent commit will move this into
defaults package. In future this can be made configurable and potentially we
might support multiple prefixes.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Move default nat46x64 prefix into defaults package

Refactor the code such that we can easily expose it as a config knob and
have the underlying code be generic.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

oblazek commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,649 @@
+maximum efficiency.
+
+The vast majority of modern network drivers for 10/40/100G and beyond support
+XDP today. If you are unsure whether your network driver supports native XDP,
```
perhaps we could suggest a [link](https://github.com/iovisor/bcc/blob/master/docs/kernel-versions.md#xdp) with supported drivers?

borkmann replied:
we also have a list in our kube-proxy-free guide. will think about it for a follow-up.

oblazek commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,649 @@
+
+Cilium exposes its programmable API under ``unix:///var/run/cilium/cilium.sock``.
+Cilium's `service command-line interface <https://github.com/cilium/cilium/blob/master/cilium/cmd/service_update.go>`_
+communicates to the agent through through this API. For simplicity, we use the
```
```suggestion
communicates to the agent through this API. For simplicity, we use the
```

oblazek commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,649 @@
+
+.. parsed-literal::
+
+  docker exec cilium-gateway cilium service update --id 1 --frontend "86.109.5.207:8080" --backends "1.1.1.1:80,1.0.0.1:80" --k8s-external
```
is the `--k8s-external` prefered in any way compared to `--k8s-nodeport`?

borkmann replied:
In this case both have the same semantics.

borkmann replied:
Perhaps if nothing is specified, we just default to `--k8s-external`, which would also simplify the doc. Will check..

oblazek commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,649 @@
+node is redirecting the request to the remote backend on its behalf by performing
+SNAT. This does not require any additional MTU changes. The cost is that replies
+from the backend need to make the extra hop back to the load-balancer node to
+perform the reverse SNAT translation there before returning the packet directly
```
```suggestion
perform the reverse SNAT translation before returning the packet directly
```
wdyt ? sounds better to me

oblazek commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,668 @@
+are still kept in-tact until a backend is removed from the given service entirely. Once
+the backend is removed from the service, then (still) ongoing traffic will be dropped.
+
+The backend state is presented in the servie dump, and can be one of ``active`` (default),
```
```suggestion
The backend state is presented in the service dump, and can be one of ``active`` (default),
```

aspsk commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1278 @@
+replace legacy IPVS-based load-balancers, and many more.
+
+Cilium provides a programmable API for its agent which is exposed through a
+unix domain socket on the local node such that it can be integrated into
```
_a_ local node?

aspsk commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1278 @@
+Quick-Start
+===========
+
+Given the Cilium Standalone Gateway can be operated outside of a Kubernetes
```
Should we just call it "Cilium Gateway"?

aspsk commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1278 @@
+===========
+
+Given the Cilium Standalone Gateway can be operated outside of a Kubernetes
+cluster, this guide provides a minimal Docker-based step-by-step instructions.
```
`a minimal ... instructions` doesn't compile: maybe you wanted to say something like "a minimal ... setup"?

aspsk commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1278 @@
+utilize Maglev consistent hashing, and the networking device for the
+load-balancer in this case is called ``bond0``. Multi-device setups are
+supported as well, and can be specified via ``--devices=bond0,eth0`` or
+through a device name regex. Furthermore, we enable XDP acceleration for
```
Not sure what `Furthermore` references to, maybe just "We enable XDP..."

aspsk replied:
And move this sentence to the next paragraph which discusses XDP

aspsk commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1281 @@
+   [...]
+
+As an optional next step, we will create a simple service entry for testing purpose.
+Then we validate that Cilium installed the service correctly.
```
will validate

aspsk reviewed (commented):
Some minor changes (didn't finish reading yet, more to come on Monday)

brb commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1281 @@
+
+.. parsed-literal::
+
+   docker run --name cilium-gateway -td \\
```
Nit: s/cilium-gateway/cilium-loadbalancer/

borkmann replied:
I see gateway term as a superset of both load-balancer and nat46/64 proxy.

brb commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1281 @@
+     -v /lib/modules:/lib/modules \\
+     --privileged=true \\
+     --network=host \\
+     --cap-add=NET_ADMIN \\
```
Why this one is needed? In the [PR](https://github.com/cilium/cilium/pull/22653) I've dropped all caps. Or is there going to be some level=warn log msg?

borkmann replied:
Ah nice, I'll double check on Mon then, would be nice if we can drop it indeed.

borkmann replied:
works indeed, fixed

ti-mo commented on pkg/option/config.go:
```diff
@@ -2591,13 +2598,37 @@ func (c *DaemonConfig) validateIPv6ClusterAllocCIDR() error {
+		return err
+	}
+
+	if cidr == nil {
```
This cannot be true if `err == nil` and can also be removed in `validateIPv6ClusterAllocCIDR`.

ti-mo commented on pkg/option/config.go:
```diff
@@ -1374,6 +1374,13 @@ type DaemonConfig struct {
+	IPv6NAT46x64CIDR string
+
+	// IPv6NAT46x64CIDRBase is derived from IPv6NAT46x64CIDR and contains
+	// the CIDR without the mask.
```
Technically, CIDR is the `<addr/prefix>/<mask>` notation, so once that's been converted from string into binary, I think we should only use the term 'prefix' in case of ipv6. `the CIDR without the mask` may be a bit ambiguous, maybe `the IPv6 prefix with the masked bits zeroed out`?

ti-mo commented on pkg/datapath/linux/config/config.go:
```diff
@@ -325,8 +325,17 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
-			cDefinesMap["ENABLE_NAT_46X64_STATELESS"] = "1"
+		if option.Config.EnableNat46X64Gateway {
+			cDefinesMap["ENABLE_NAT_46X64_GATEWAY"] = "1"
+			cDefinesMap["NAT_46X64_PREFIX_0"] = fmt.Sprintf("%d", int(option.Config.IPv6NAT46x64CIDRBase[0]))
```
This may be stating the obvious, but this depends completely on `net.IP` representing its address in net byte order. Could the meaning of `NAT_46X64_PREFIX_X` (at least the byteorder) be documented for when we move to `netip`?

@tklauser Can `netip.Addr.AsSlice()` be used to implement the equivalent here?

ti-mo replied:
The int conversion can be dropped, `byte` is `uint8`.
```suggestion
			cDefinesMap["NAT_46X64_PREFIX_0"] = fmt.Sprintf("%d", option.Config.IPv6NAT46x64CIDRBase[0])
```

tklauser replied:
> @tklauser Can `netip.Addr.AsSlice()` be used to implement the equivalent here?

Yes, as Sebastian also [noted](https://github.com/cilium/cilium/pull/22421/files#r1045964692) `IPv6NAT46x64CIDRBase` should ideally be of `netip.Addr` type and we could then use `netip.Addr.AsSlice()` as Timo suggests.

ti-mo commented on pkg/option/config.go:
```diff
@@ -2591,13 +2598,37 @@ func (c *DaemonConfig) validateIPv6ClusterAllocCIDR() error {
+	}
+
+	if ones, _ := cidr.Mask.Size(); ones != 96 {
+		return fmt.Errorf("CIDR length must be /96")
```
Technically, this prefix length, not CIDR length.

ti-mo reviewed (approved):
Thanks, left a few nits to ease future work.

gandro commented on pkg/option/config.go:
```diff
@@ -2591,13 +2598,37 @@ func (c *DaemonConfig) validateIPv6ClusterAllocCIDR() error {
 }
 
+func (c *DaemonConfig) validateIPv6NAT46x64CIDR() error {
+	ip, cidr, err := net.ParseCIDR(c.IPv6NAT46x64CIDR)
```
optional nit: I understand this is copied over from `validateIPv6ClusterAllocCIDR`, but consider using `netip.ParsePrefix` for new code.

gandro reviewed (approved):
My CODEOWNERs look good to me.

tklauser reviewed (approved):
Looks good for my codeowners.

aspsk commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1470 @@
+  2022-12-12T12:39:15Z Output: /var/run/cilium/pcaps/hubble_1670848754_1823804162_c3-small-x86-01.pcap
+
+Another example for recording TCP or UDP-based traffic with a source CIDR
+of ``10.4.0.0/16`` and any port, a destrination CIDR of ``1.1.1.1/32`` and
```
typo: destrination -> destination

aspsk commented on Documentation/network/standalone.rst:
```diff
@@ -243,7 +246,7 @@ And last but not least deleted through its identifier:
 
 Each of these operations communicate to the agent through its programmable API
-which for third party integrations can be used directly.
+which can be used directly, for eample for third party integrations.
```
eample -> example

aspsk commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1493 @@
+
+By default, Hubble stores recorded PCAPs for post-analysis under ``/var/run/cilium/pcaps/``.
+
+Example for recording all TCP-based traffic for the node for a time of one second:
```
One comment here is that hubble won't actually dump _all_ the traffic inside pcap: if a `cilium_capture_in` isn't matched by `cilium_capture_out`. One case is when traffic is dropped by the host. The cilium event will be sent in any case by `cilium_capture_in`, so a custom control plane can dump such traffic as well

borkmann replied:
makes sense, added comment

aspsk reviewed (commented):
Just two comments (more to come later)

borkmann committed:
cilium: Remove unneeded CIDR nil check

This cannot be true if err == nil and can also be removed in
validateIPv6ClusterAllocCIDR.

Also, s/CIDR length must be/prefix length must be/.

Suggested-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Use netip.Addr for IPv6NAT46x64CIDRBase

For new code, preference is to use net/netip package.

Suggested-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -325,8 +330,13 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
-			cDefinesMap["ENABLE_NAT_46X64_STATELESS"] = "1"
+		if option.Config.EnableNat46X64Gateway {
+			cDefinesMap["ENABLE_NAT_46X64_GATEWAY"] = "1"
+			base := option.Config.IPv6NAT46x64CIDRBase.AsSlice()
+			cDefinesMap["NAT_46X64_PREFIX_0"] = fmt.Sprintf("%d", base[0])
+			cDefinesMap["NAT_46X64_PREFIX_1"] = fmt.Sprintf("%d", base[1])
+			cDefinesMap["NAT_46X64_PREFIX_2"] = fmt.Sprintf("%d", base[2])
+			cDefinesMap["NAT_46X64_PREFIX_3"] = fmt.Sprintf("%d", base[3])
```
Is there a reason not to use `defineIPv6()` like the other IPv6 material? Seems like this is a mask / prefix but other than that it should follow pretty much the same pattern. `DEFINE_IPV6` on the datapath side would handle the `_0`, `_1`, etc. suffixes.

joestringer commented on bpf/node_config.h:
```diff
@@ -252,3 +252,10 @@ return false;
 #define CIDR_IDENTITY_RANGE_START ((1 << 24) + 1)
 #define CIDR_IDENTITY_RANGE_END   ((1 << 24) + (1<<16) - 1)
+
+#ifndef NAT_46X64_PREFIX_0
```
`DEFINE_IPV6(...)`?

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+- Load-balancer for legacy VM-based workloads,
+- As a gateway in front of IPv6 single stack Kubernetes clusters to allow for
+  interconnection with IPv4,
+- As a load-balancer for kubeapi-server,
```
nit, I think this will satisfy the spellchecker too and avoid needing to spellcheck `kubeapi` which is a less common abbreviation:
```suggestion
- As a load-balancer for kube-apiserver,
```

joestringer replied:
Switching to the content, is this a common need to expose kube-apiserver outside the cluster? Seems like something that most users would want to avoid due to security concerns 🤔

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+``/sys/fs/bpf``. If this is not the case already, a mount unit file can be set
+up under ``/etc/systemd/system/sys-fs-bpf.mount`` ...
+
+.. parsed-literal::
```
`parsed-literal` enables additional preprocessing such as substituting links in the output based on predefined macros like `\ |SCM_WEB|`. This should ideally be formatted with some systemd syntax highlighting, but I'd guess that doesn't exist in sphinx.

Easiest path is just to use plain formatting of a text block:

```suggestion
::
```

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+  The above step is only required for distributions with older systemd versions.
+  For most modern distributions this step is not required as systemd will
+  automatically mount the BPF file system at its default location. See also
+  `GH-10955 <https://github.com/cilium/cilium/issues/10955>`_ for details.
```
I wonder if we should just put this step in some other requirements file that you have to click through, then simplify this doc to just say "Ensure that the BPF filesystem is mounted. This is enabled by default in all modern distributions, but if you are unsure then see [reference to the other page]".

This would allow a bunch of configuration output here to be hidden which seems ideal given that 90% of users shouldn't need it.

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+functionality for both IPv4 and IPv6, with Maglev consistent hashing as the
+load-balancing algorithm. In our case, the networking device for the load-balancer
+is called ``bond0``. Multi-device setups are supported as well, and can be
+specified via ``--devices=bond0,eth0`` or through a device name regex.
```
Is it really regex? I thought it's that odd iptables-esque syntax like `eth+`

borkmann replied:
Will clarify, I think it's a simplified form, but will double check.

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+
+.. parsed-literal::
+
+   docker exec cilium-gateway cilium status --verbose
```
This will syntax-highlight the commands correctly and avoid the sphinx literal substitution:

```suggestion
.. code-block:: shell-session

   $ docker exec cilium-gateway cilium status --verbose
```

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+As an optional next step, we will create a simple service entry for testing purpose.
+Then we will validate that Cilium installed the service correctly.
+
+Cilium exposes its programmable API under ``unix:///var/run/cilium/cilium.sock``.
```
We probably want to drop a note here somewhere that the API is subject to change? I don't think we're currently enforcing API stability on these changes.

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+
+After initial deployment the load-balancer service table is empty:
+
+.. parsed-literal::
```
More `parsed-literal`, typically these should be `.. code-block:: shell-session` unless there is actually a variable substitution occurring. I won't point out remaining cases of this for the rest of the review.

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+
+.. parsed-literal::
+
+  docker exec cilium-gateway cilium service list
```
Incidentally I'm not sure that the indentation is quite right here? Typically we suggest 4-space indentation for all of these sphinx snippets. I think that for the first layer of nesting it might not matter, but if you ever need to double-nest sphinx output then you may need to follow the 4-space indentation consistently to get the hunks to format right.

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+  1    192.168.160.3:8080   ExternalIPs    1 => 192.168.0.3:4444
+
+With the default ``--bpf-lb-dsr-l4-xlate=frontend``, the inbound and outbound packet look
+as follows:
```
```suggestion
With the default ``--bpf-lb-dsr-l4-xlate=frontend``, the inbound and outbound packet look
as follows in ``cilium monitor`` output:
```

(or you could just put a cilium monitor example command into the snippet)

borkmann replied:
it was actually tcpdump :)

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+load-balancer responds with an IPv4 ICMP ``destination unreachable`` message with sub-type
+``fragmentation needed``, so that clients are able to cache this path information and
+to adjust their packet sizes for future transmissions. The IPv6 counterpart emits an
+ICMPv6 ``Packet Too Big`` message back to the sender. Both is auto-enabled under XDP
```
```suggestion
ICMPv6 ``Packet Too Big`` message back to the sender. Both are auto-enabled under XDP
```

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+.. parsed-literal::
+
+   SEED=$(head -c12 /dev/urandom | base64 -w0)
+   echo $SEED
```
nit, we have the debate back and forth sometimes, but for a case like this I think it's helpful to get the `shell-session` highlight for the `SEED` and `echo` lines (since those are the ones you type into the terminal) then observe the output of the commands. Like so:
```suggestion
.. code-block:: shell-session
   $ SEED=$(head -c12 /dev/urandom | base64 -w0)
   $ echo $SEED
```

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+- ``active`` into ``terminating``, ``quarantined`` or ``maintenance``
+- ``quarantined`` into ``active`` or ``terminating``
+- ``maintenance`` into ``active``
+- ``terminating`` is a final state
```
Might be nice to represent this as a graph, but 👍

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+  1    86.109.5.207:8080   ExternalIPs    1 => 1.1.1.1:80 (active)
+                                          2 => 1.1.1.2:80 (maintenance)
+
+The backend state is global, meaning, if a backend IP:port is part of multiple services,
```
"global" can have implications around multi-cluster, do we want to use that terminology here? (I realise standalone gateway is a bit of a different beast, but I figure I should ask in case it does overlap / confuse).

borkmann replied:
changed into:

The backend state is global for the load-balancer node, meaning, if a backend IP:port is part
of multiple services, then all of them are updated accordingly:

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+    * Upon Cilium restart the backend weights are preserved as-is, however, the
+      displayed weights value in the ``cilium service list`` json/yaml dump is
+      currently not preserved and will display ``1``. See further details
+      in `GH-18306 <https://github.com/cilium/cilium/pull/18306>`_. This does
```
```suggestion
      in :gh-issue:`18306`. This does
```

joestringer replied:
^ There's probably more cases of this, we use the formatter to get consistent output when referring to github issues across the docs.

joestringer commented on Documentation/network/standalone.rst:
```diff
@@ -0,0 +1,1610 @@
+    * PMTU discovery with ICMPv4/ICMPv6 replies is currently only supported
+      under XDP mode.
+    * Cilium's eBPF load-balancer does not support the SCTP transport protocol.
+      Only TCP and UDP is supported as a transport for services at this point.
```
Could probably also reference the sctp gh issue here?

joestringer reviewed (changes requested):
Mostly just doing a pass at the docs-structure / sphinx usage, all pretty minor stuff. Thanks for the nice writeup! I didn't look closely at the bpf pieces since it seemed like others already reviewed those in detail.
